### ISSUE TITLE
lint matchable `if let` for pattern

### ISSUE THREAD

**Issue by dswij** [rust-lang/rust-clippy#9221]
### What it does

Checks for `if let` for `pattern`s that can be re-written with `matches!`

This can be an improvement to [`equatable_if_let`](https://rust-lang.github.io/rust-clippy/master/#equatable_if_let) lint.

For background see https://github.com/rust-lang/rust-clippy/pull/9102#discussion_r924363140

### Advantage

Similar to `equatable_if_let`:

1. It reads better and has less cognitive load because equality won’t cause binding.
2. It avoids a [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions).
3. Equality is a simple bool expression and can be merged with && and || and reuse if blocks

### Example

```rust
#![allow(dead_code)]
enum Foo {
    Bar,
    Baz,
}

fn test(foo: Foo) {
    if let Foo::Baz = foo {
        println!("Foo::Baz");
    }
}
```

can be written as

```rust
#![allow(dead_code)]
enum Foo {
    Bar,
    Baz,
}

fn test(foo: Foo) {
    if matches!(foo, Foo::Baz) {
        println!("Foo::Baz");
    }
}
```

**Comment by dswij** [comment]
We can use `equatable_if_let` as a starting point, this should be a good first issue

**Comment by dswij** [comment]
Also, the logic would also be similar to [`match_like_matches_macro`](https://rust-lang.github.io/rust-clippy/master/#match_like_matches_macro).

**Comment by nahuakang** [comment]
@dswij 👋 May I ask if you meant that we should add the new linting within `equatable_if_let`, i.e. expanding the logic of the existing `equatable_if_let` lint to include this new pattern? Thanks :)

**Comment by dswij** [comment]
I haven't taken a deep look into it, but if we plan to add this to a new lint, we can factor out reusable logic from `equatable_if_let`. Alternatively, we can just add this case to be linted by `equatable_if_let` and update the lint's docs.

**Comment by nahuakang** [comment]
Thanks. I'll take a deeper look at the 2 lints that you've written down and circle back when I have a clearer picture :)

**Comment by nahuakang** [comment]
@dswij It seems that `equatable_if_let` [already takes care of the case](https://github.com/rust-lang/rust-clippy/blob/5735a3bef6db84156d78d05fdd43e6332687cf2c/tests/ui/equatable_if_let.stderr#L39-L62) you mentioned (I wonder if dogfood would have caught it). In this case, do we prefer: `if foo == Foo::Baz` or `if matches!(foo, Foo::Baz)`?

**Comment by dswij** [comment]
@nahuakang 

it seems like it won't fire if `PartialEq` is not implemented. See ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2d6cadfff9f95c52799ed27409769df6)). The test cases for this lint seems to agree.

Makes sense since if `PartialEq` is not implemented, then you can't suggest `if foo == Foo::Baz`

So I think this is more of an improvement - if it doesn't implement `PartialEq`, then we can suggest using `matches` instead.

**Comment by nahuakang** [comment]
@dswij Thanks for the tip. I was trying to understand this lint but had some trouble yesterday. Does your suggestion mean that if [is_structural_partial_eq](https://github.com/rust-lang/rust-clippy/blob/5735a3bef6db84156d78d05fdd43e6332687cf2c/clippy_lints/src/equatable_if_let.rs#L76) returns `false`, we could consider suggesting `matches`?

**Comment by nahuakang** [comment]
Will do! I'll start adding some new test cases and do the improvement. Maybe by the time of review we'd think of more edge cases.

**Comment by nahuakang** [comment]
@dswij Feel free to take a look at [the PR](https://github.com/rust-lang/rust-clippy/pull/9368) :) I'm still not familiar with things on this level so might miss some edge cases.

**Comment by ivan-aksamentov** [comment]
@nahuakang Excuse me if this is a wrong place to ask, but after this change to the `equatable_if_let`, it turns conditions like

```rust
if let (None, None) = (&self.x, &self.y) {
```

into

```rust
if matches!((&self.x, &self.y), (None, None)) {
```

which I would consider worse.

Otherwise `equatable_if_let` is a fantastic, helpful lint, but I think this `matches!()` addition is bolted into a wrong place. Is it not too late to extract the `matches!()` part of it into a separate lint (which I would disable) and to roll back the previous version of the `equatable_if_let`?

How do I properly request or open a discussion for this?
